ethcore/src/client/client.rs

```diff
@@ -1785,26 +1785,49 @@ impl BlockChainClient for Client {
 	}
 
 	fn transaction_receipt(&self, id: TransactionId) -> Option<LocalizedReceipt> {
+		// NOTE Don't use block_receipts here for performance reasons
+		let address = self.transaction_address(id)?;
+		let hash = address.block_hash;
 		let chain = self.chain.read();
-		self.transaction_address(id)
-			.and_then(|address| chain.block_number(&address.block_hash).and_then(|block_number| {
-				let transaction = chain.block_body(&address.block_hash)
-					.and_then(|body| body.view().localized_transaction_at(&address.block_hash, block_number, address.index));
-
-				let previous_receipts = (0..address.index + 1)
-					.map(|index| {
-						let mut address = address.clone();
-						address.index = index;
-						chain.transaction_receipt(&address)
-					})
-					.collect();
-				match (transaction, previous_receipts) {
-					(Some(transaction), Some(previous_receipts)) => {
-						Some(transaction_receipt(self.engine().machine(), transaction, previous_receipts))
-					},
-					_ => None,
-				}
-			}))
+		let number = chain.block_number(&hash)?;
+		let body = chain.block_body(&hash)?;
+		let mut receipts = chain.block_receipts(&hash)?.receipts;
+		receipts.truncate(address.index + 1);
+
+		let transaction = body.view().localized_transaction_at(&hash, number, address.index)?;
+		let receipt = receipts.pop()?;
+		let gas_used = receipts.last().map_or_else(|| 0.into(), |r| r.gas_used);
+		let no_of_logs = receipts.into_iter().map(|receipt| receipt.logs.len()).sum::<usize>();
```

Comment:
I don't understand these 3 lines. why do we pop a receipt and then take a `gas_used` from the one before it?

Reply:
So to actually generate a `LocalizedReceipt` we need 3 things:
1. Actual receipt
2. Gas used by that transaction
3. Number of logs generate prior to this transaction (to calculate log_index_in_block).

So the logic is as follows:
1. Fetch all receipts from 0 to `index` (inclusive)
2. Take the last one (`receipts.pop()`; point 1.)
3. Take the previous to last one to calculate (point 2.) (receipts only store cumulative gas, so we calculate `current.gas_used - previous.gas_used` to figure out what `current` gas usage of transaction was)
4. Use all previous receipts to calculate total number of logs. (point 3.)